<div style="text-align: center;">
<h2>INFSCI 2915 Foundations- Machine Learning - Spring 2018 </h2>
<h1 style="font-size: 250%;">Assignment #3</h1>
<h3>Issued Monday, 3/26/2018; Due Monday, 11:59pm, 4/09/2018</h3>
<h3>Total points: 100 </h3>
</div>

In [35]:
# Type in your information in the double quotes
firstName = "Junxi"
lastName = "Huang"
pittID = "juh80"

<h3>  Problem #1.Linear Discriminant Analysis (LDA)and Quadratic Discriminant Analysis(QDA)      <br>[30 points]</h3> 
 Writing a code is not required for this problem.

<h4> Problem #1-1</h4>  <br>
Assume we have K classes to be classified with one feature $(x)$. The prior probability of
class $k$ is $𝜋_{k} = 𝑃(𝑌 = 𝑘)$. Assume that the feature in class k has Gaussian distribution of
mean $μ_{k}$ and variance $σ^2 (𝒩(μ,𝜎^{2}))$.The variance is the same for all classes. 
Prove that the Bayes’ classifier (that chooses class k with largest $𝑃(𝑌 = 𝑘|𝑥))$ is equivalent to assigning an observation to the class for which the discriminant function $𝛿_{k}(x)$ is
maximized, where 
\begin{array} \\
𝛿_{k}(x) = x\frac{\mu _{k}}{\sigma ^{2}}- \frac{\mu_{2}^{k}}{2\sigma ^{2}}+ log(\pi _{k})
\end{array}
<br> What is the name of this classifier?


### Answer(proof on picture):

Linear Discriminant Analysis


In [2]:
from IPython.display import Image
Image(filename='homework3.jpg', width=300, height=200)

IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.


<h4> Problem #1-2</h4>  <br>
Extend **Problem #1-1** to include **p** features. With features from each class drawn from a
Gaussian distribution with mean vector $μ_{k}$ and covariance matrix $Σ_{k}$ (which is now
different for each class). Find the discriminant function that maximizes **𝑃(𝑌 = 𝑘|𝑥)**. Is
the relationship with the feature vector **x** linear?<br> What is this classifier?


### Answer1-2
we get discriminate functions that are quadratic with x.
name is:  
Quadratic Discriminant Analysis(QDA)

<h4> Problem #1-3</h4>  <br>
- Explain Bias-variance trade-off in choosing between LDA and QDA

### Answer 1-3
LDA is simple which has lower variance, but QDA has lower bias

<h3>  Problem #2. Support Vector Machines  [20 points]</h3> 
Writing a code is not required for this problem.

<h4> Problem #2-1 Answer the following quesionts </h4> 

- Describe limitations of using Maximal Margin Classifier 
- What is the difference between Support Vector Classifier and Maximal Margin Classifier
- Explain Bias-variance trade-off that occurs when we choose large and small values for the Slack Variables


### Answer 2.1:
1.Maximal Margin Classifier is sensible to training data and classes may not be perfectly seperable by a hyperline

2.Support vector classifier allows some training observations to be in the
incorrect side of the hyperplane (decision boundary)

3.if we choose large values for Slack Variables, it means the modle will more tolerance to validation with wide margin, which means large bias and small variance.                                                                                         
if we choose large values for Slack Variables, it means the modle will less tolerance to validation with narrow margin, which means small bias and large variance.


<h4> Problem #2-2</h4><br>
How does the Radial Basis Function Kernel in SVM measure the similarity between a test point and a training example? Discuss the impact of choosing a large RBF parameter **𝛾**
on the learning algorithm.

### Answer 2.2:
It computes the squared Euclidean distance between the observation point and training point. Small distance means more similarity, otherwise means less similarity.   
A large RBF parameter 𝛾 (small variance) may overfit as it becomes more local

<h3>  Problem #3 Classification Performance Evaluation and Cross validation [30 points]</h3> 


<h4> Problem #3-1</h4> <br>
In a fraud detection system, a QDA classifier’s confusion matrix is found to be:

|        |Predicted Class - Not fraud| Predicted Class - fraud|
|:--:|:-------------------------------:|
|Actual class – Not fraud|1000|20|
|Actual class – fraud|30|5|

- Evaluate the overall error rate and the accuracy<br>
- Evaluate the precision and the recall <br>
- Is dataset balanced?
- Comment on the results.

In [4]:
a=1/7
a

0.14285714285714285

### Answer 3-1
1.The accuracy is :(1000+5)/(1000+5+20+30)=95.26%                                                                               
  The error rate is (1-accuracy):(1-0.9526)=0.0474%
  
2.percision is true positive/all predict positive: 5/(20+5)=20%                                                                 
  recall is true positive/ all actual positive: 5/(30+5)=14.29%
  
3.the dataset is not balanced, because samples of fraud class is much less than samples of not fraud class.

4.The performance of this system on detect fraud class are bad.

<h4> Problem #3-2</h4> <br>
Assuming we want to lower the misclassification of fraud. How can you modify the classifier to do better classification using information from confusion matrix? 

### Answer 3-3
1.Expand the data set. 

2.Resampling, for small data, apply oversampling, for big data, apply undersampling. 

3.Attempt to generate artificial data samples, like SMOTH method. 

4.try different classification algorithom, like decision tree. 

5.add punishment term to model. when the model classify the class of samll samples, add another punishment term model so that the model can more focus on the class of samll samples. 

<h4> Problem #3-3 Cross validation, SVM </h4> <br>
In this exercise, we will use SVM  for breast cancer classification (malignant or benign).<br> 
Use a code below to download the dataset:


In [21]:
from sklearn.datasets import load_breast_cancer
dataset = load_breast_cancer()
print(dataset.keys())


dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names'])


Follow steps to answer questions.
- Scale the features with MinMaxScaler
- Split breast cancer dataset into two, test dataset and train dataset
- Find best SVM classifier model. Try values of C=[0.001, 0.1, 1, 10, 1000] and Gamma = [0.001, 0.1, 1, 10, 1000]. 
- Use 3-fold cross validation to find the best parameters (using all possible combinations of these values for C and gamma).
- Report your parameters
- Report accuracy, confusion matrix, precision, and recall (use Test dataset, SVM classifier model with best parameters)
- Comment your results

In [33]:
data = dataset['data']
target = dataset['target']
print(data[0])
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
scaler.fit(data)
data = scaler.transform(data)
print(data[0])

[  1.79900000e+01   1.03800000e+01   1.22800000e+02   1.00100000e+03
   1.18400000e-01   2.77600000e-01   3.00100000e-01   1.47100000e-01
   2.41900000e-01   7.87100000e-02   1.09500000e+00   9.05300000e-01
   8.58900000e+00   1.53400000e+02   6.39900000e-03   4.90400000e-02
   5.37300000e-02   1.58700000e-02   3.00300000e-02   6.19300000e-03
   2.53800000e+01   1.73300000e+01   1.84600000e+02   2.01900000e+03
   1.62200000e-01   6.65600000e-01   7.11900000e-01   2.65400000e-01
   4.60100000e-01   1.18900000e-01]
[ 0.52103744  0.0226581   0.54598853  0.36373277  0.59375282  0.7920373
  0.70313964  0.73111332  0.68636364  0.60551811  0.35614702  0.12046941
  0.3690336   0.27381126  0.15929565  0.35139844  0.13568182  0.30062512
  0.31164518  0.18304244  0.62077552  0.14152452  0.66831017  0.45069799
  0.60113584  0.61929156  0.56861022  0.91202749  0.59846245  0.41886396]


In [34]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(data, target, test_size=0.33, random_state=0)
print('X_tain shape:',X_train.shape)
print('X_test shape:',X_test.shape)
print('y_train shape:',y_train.shape)
print('y_test shape:',y_test.shape)

X_tain shape: (381, 30)
X_test shape: (188, 30)
y_train shape: (381,)
y_test shape: (188,)


In [39]:
from sklearn.svm import SVC
max_method=''
max_gamma=0.001
max_score=0
for i in [0.001, 0.1, 1, 10, 1000] :
    for x in ['rbf', 'linear', 'poly','sigmoid']:
        svmModel=SVC(kernel=x, gamma=i, C=i).fit(X_train,y_train)
        score=svmModel.score(X_test,y_test)
        if(score>max_score):
            max_method=x
            max_gamma=i
            max_score=score
        print('x=',x,'   gamma and c:',i,'   score:',score)
print('\n','best\n','max_method:',max_method,'    gamma:',max_gamma,'   score:',max_score)

x= rbf    gamma and c: 0.001    score: 0.643617021277
x= linear    gamma and c: 0.001    score: 0.643617021277
x= poly    gamma and c: 0.001    score: 0.643617021277
x= sigmoid    gamma and c: 0.001    score: 0.643617021277
x= rbf    gamma and c: 0.1    score: 0.909574468085
x= linear    gamma and c: 0.1    score: 0.941489361702
x= poly    gamma and c: 0.1    score: 0.696808510638
x= sigmoid    gamma and c: 0.1    score: 0.803191489362
x= rbf    gamma and c: 1    score: 0.978723404255
x= linear    gamma and c: 1    score: 0.973404255319
x= poly    gamma and c: 1    score: 0.973404255319
x= sigmoid    gamma and c: 1    score: 0.303191489362
x= rbf    gamma and c: 10    score: 0.941489361702
x= linear    gamma and c: 10    score: 0.973404255319
x= poly    gamma and c: 10    score: 0.925531914894
x= sigmoid    gamma and c: 10    score: 0.643617021277
x= rbf    gamma and c: 1000    score: 0.643617021277
x= linear    gamma and c: 1000    score: 0.941489361702
x= poly    gamma and c: 1000   

In [46]:
#Use 3-fold cross validation to find the best parameters (using all possible combinations of these values for C and gamma).
#Report your parameters

from sklearn.model_selection import cross_val_score
max_method=''
max_gamma=0.001
max_score=0
for i in [0.001, 0.1, 1, 10, 1000] :
    for x in ['rbf', 'linear', 'poly','sigmoid']:
       for j in [0.001, 0.1, 1, 10, 1000] :
            svmModel=SVC(kernel=x, gamma=i, C=j)
            score = cross_val_score(svmModel, data, target, cv=3)
            if(score.mean()>max_score):
                max_method=x
                max_gamma=i
                max_c=j
                max_score=score.mean()
            print('x=',x,'   gamma:',i,'   c:',j,'   score:',score.mean())
print('\n','best\n','max_method:',max_method,'    gamma:',max_gamma,'      c:',max_c,'   score:',max_score)

x= rbf    gamma: 0.001    c: 0.001    score: 0.627420402859
x= rbf    gamma: 0.001    c: 0.1    score: 0.627420402859
x= rbf    gamma: 0.001    c: 1    score: 0.627420402859
x= rbf    gamma: 0.001    c: 10    score: 0.906943284136
x= rbf    gamma: 0.001    c: 1000    score: 0.971883412234
x= linear    gamma: 0.001    c: 0.001    score: 0.627420402859
x= linear    gamma: 0.001    c: 0.1    score: 0.945558340295
x= linear    gamma: 0.001    c: 1    score: 0.9718926947
x= linear    gamma: 0.001    c: 10    score: 0.963092917479
x= linear    gamma: 0.001    c: 1000    score: 0.94905782976
x= poly    gamma: 0.001    c: 0.001    score: 0.627420402859
x= poly    gamma: 0.001    c: 0.1    score: 0.627420402859
x= poly    gamma: 0.001    c: 1    score: 0.627420402859
x= poly    gamma: 0.001    c: 10    score: 0.627420402859
x= poly    gamma: 0.001    c: 1000    score: 0.627420402859
x= sigmoid    gamma: 0.001    c: 0.001    score: 0.627420402859
x= sigmoid    gamma: 0.001    c: 0.1    score: 0.

In [48]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import recall_score, precision_score

svcModel = SVC(C = 1.0,gamma = 1.0).fit(X_train,y_train)
prediction = svcModel.predict(X_test)
score = svcModel.score(X_test, y_test)
print("score: ",score)
cm = confusion_matrix(y_test, prediction).tolist()
print("confusion matrix: \n",cm)
recall = recall_score(y_test, prediction)
print("recall: ",recall)
precision = precision_score(y_test, prediction)
print("precision: ",precision)

score:  0.978723404255
confusion matrix: 
 [[64, 3], [1, 120]]
recall:  0.99173553719
precision:  0.975609756098


### Answer 3-3:
the best parameter is: max_method: rbf     gamma: 1       c: 1    score: 0.97716513506

<h4> Problem #3-4 Cross validation, Logistic regression </h4>

- Repeat **Problem #3-3** but instead of SVM find best Logistic regression classifier model. Try values of C= [0.001,0.01,1,10,100, 1000] and Penalty = ["l1","l2"]


In [58]:
from sklearn.linear_model import LogisticRegression
C= [0.001,0.01,1,10,100, 1000]

Penalty='11'
max_c=0.001
max_score=0

print("penalty is l1:")
for i in C :
    log = LogisticRegression(penalty = 'l1',C = i)
    score = cross_val_score(log, data, target, cv=3)
    print(score.mean())
    if(score.mean()>max_score):
            Penalty='11'
            max_c=i
            max_score=score.mean()

print("penalty is l2:")
for i in C :
    log = LogisticRegression(penalty = 'l2',C = i)
    score = cross_val_score(log, data, target, cv=3)
    print(score.mean())
    if(score.mean()>max_score):
            Penalty='12'
            max_c=i
            max_score=score.mean()
            
print('\n','best\n','penalty:',Penalty,'      c:',max_c,'   score:',max_score)

penalty is l1:
0.372579597141
0.372579597141
0.961338531514
0.961338531514
0.959593428014
0.949067112225
penalty is l2:
0.688981713543
0.864800891117
0.954339552585
0.9718926947
0.971883412234
0.963102199944

 best
 penalty: 12       c: 10    score: 0.9718926947


In [57]:
log = LogisticRegression(penalty = 'l2',C =10 ).fit(X_train,y_train)
prediction = log.predict(X_test)
score = log.score(X_test, y_test)
print("score: ",score)
cm = confusion_matrix(y_test, prediction).tolist()
print("confusion matrix: \n",cm)
recall = recall_score(y_test, prediction)
print("recall: ",recall)
precision = precision_score(y_test, prediction)
print("precision: ",precision)

score:  0.973404255319
confusion matrix: 
 [[63, 4], [1, 120]]
recall:  0.99173553719
precision:  0.967741935484


<h3>  Problem #4  Neural Networks [20 points]</h3> 
In this exercise we will clasify Iris species (Setosa, Versicolor, Virginica) using Neural Networks.<br>
Use a code below to download the dataset:

In [47]:
from sklearn.datasets import load_iris
dataset=load_iris()
print(dataset.keys())

dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names'])


Follow steps to answer questions.
- Scale the feautures with MinMaxScaler, then use sklearn MLPClassifier. 
- Build a model that has two hidden layers, the first layer has 10 neurons and second layer has 5 neurons. 
- Use 'relu' activation function, and set the regularization parameter alpha=0.5. 
- Set the random_state=0 for splitting and building all models.


<h4> Problem #4-1 </h4>
- Use gradient descent to solve the optimization  problem (i.e. get the weights), and choose random_state=0 (which corresponds to a particular initializationo of weight values). 
- Report accuracy, confusion matrix, precision, and recall 

In [67]:
#Write your answer here
from sklearn.preprocessing import MinMaxScaler
data = dataset['data']
target = dataset['target']
print('before scale: ', data[0])

scaler = MinMaxScaler()
scaler.fit(data)
data = scaler.transform(data)
print('after scale',data[0])

before scale:  [  1.79900000e+01   1.03800000e+01   1.22800000e+02   1.00100000e+03
   1.18400000e-01   2.77600000e-01   3.00100000e-01   1.47100000e-01
   2.41900000e-01   7.87100000e-02   1.09500000e+00   9.05300000e-01
   8.58900000e+00   1.53400000e+02   6.39900000e-03   4.90400000e-02
   5.37300000e-02   1.58700000e-02   3.00300000e-02   6.19300000e-03
   2.53800000e+01   1.73300000e+01   1.84600000e+02   2.01900000e+03
   1.62200000e-01   6.65600000e-01   7.11900000e-01   2.65400000e-01
   4.60100000e-01   1.18900000e-01]
after scale [ 0.52103744  0.0226581   0.54598853  0.36373277  0.59375282  0.7920373
  0.70313964  0.73111332  0.68636364  0.60551811  0.35614702  0.12046941
  0.3690336   0.27381126  0.15929565  0.35139844  0.13568182  0.30062512
  0.31164518  0.18304244  0.62077552  0.14152452  0.66831017  0.45069799
  0.60113584  0.61929156  0.56861022  0.91202749  0.59846245  0.41886396]


In [68]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
X_train, X_test, y_train, y_test = train_test_split(data, target, random_state=0)

In [69]:
from sklearn.neural_network import MLPClassifier

MLPmodel = MLPClassifier(solver='sgd', activation='relu', random_state=0, hidden_layer_sizes=[10,5], alpha=0.5)
MLPmodel.fit(X_train,y_train)
prediction = MLPmodel.predict(X_test)
score = accuracy_score(y_test,prediction)
print("accuracy score: ",score)
cm = confusion_matrix(y_test, prediction).tolist()
print("confusion matrix: \n",cm)
recall = recall_score(y_test, prediction, average='macro')
print("recall: ",recall)
precision = precision_score(y_test, prediction, average='macro')
print("precision: ",precision)

accuracy score:  0.629370629371
confusion matrix: 
 [[0, 53], [0, 90]]
recall:  0.5
precision:  0.314685314685


C:\ProgramData\Anaconda3\lib\site-packages\sklearn\neural_network\multilayer_perceptron.py:564: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1135: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)


<h4> Problem #4-2 </h4>
- Repeat **Problem #4-1** but with a model that use random_state=10 to initialize the weights. 
- Report accuracy, confusion matrix, precision, and recall

In [72]:
#Write your answer here
MLPmodel = MLPClassifier(solver='sgd', activation='relu', random_state=0, max_iter = 4000,hidden_layer_sizes=[10,5], alpha=0.5)
MLPmodel.fit(X_train,y_train)
prediction = MLPmodel.predict(X_test)
score = accuracy_score(y_test,prediction)
print(score)

0.958041958042


<h4> Problem #4-3 </h4>
- Repeat **Problem #4-2** but with model that use L-BFGS (a numerical quasi-Newton method) instead of stochastic gradient descent to find the weights.
- Report accuracy, confusion matrix, precision, and recall
- Comment on results

In [71]:
#Write your answer here
MLPmodel = MLPClassifier(solver='lbfgs', activation='relu', random_state=10, hidden_layer_sizes=[10,5], alpha=0.5)
MLPmodel.fit(X_train,y_train)
prediction = MLPmodel.predict(X_test)
score = accuracy_score(y_test,prediction)
print("accuracy score: ",score)
cm = confusion_matrix(y_test, prediction).tolist()
print("confusion matrix: \n",cm)
recall = recall_score(y_test, prediction, average='macro')
print("recall: ",recall)
precision = precision_score(y_test, prediction, average='macro')
print("precision: ",precision)

accuracy score:  0.944055944056
confusion matrix: 
 [[50, 3], [5, 85]]
recall:  0.94392033543
precision:  0.9375


### Answer 4-3
       
lbfgs is more suitable for small data sets.   
also, add more hidden layer doesn't help much.   